Missing Enum constraint on JsonNumberEnumConverter<TEnum> in reference assembly

### Description

In the source code, System.Text.Json.Serialization.JsonNumberEnumConverter\<TEnum\> has a type constraint `where TEnum : struct, Enum`.  In the reference assembly however, the constraint is just `where TEnum : struct` and omits `Enum`.  Thus, if `TEnum` is a non-enum value type, the compiler doesn't notice any problem but it fails at run time.

<https://github.com/dotnet/runtime/blob/3eb129fd57994bb7ecfff6f93259bd829423cde1/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonNumberEnumConverter.cs#L17-L18>

<https://github.com/dotnet/runtime/blob/3eb129fd57994bb7ecfff6f93259bd829423cde1/src/libraries/System.Text.Json/ref/System.Text.Json.cs#L981>

### Reproduction Steps

NonEnum.csproj:

```csharp
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>
```

Program.cs:

```csharp
_ = new System.Text.Json.Serialization.JsonNumberEnumConverter<int>();
```

`dotnet run`

### Expected behavior

The compiler should report an error:

> Program.cs(1,64): error CS0315: The type 'int' cannot be used as type parameter 'TEnum' in the generic type or method 'JsonNumberEnumConverter\<TEnum\>'. There is no boxing conversion from 'int' to 'System.Enum'.
>
> The build failed. Fix the build errors and run again.

### Actual behavior

The build succeeds without any warnings, but the program fails at run time:

> Unhandled exception. System.TypeLoadException: GenericArguments[0], 'System.Int32', on 'System.Text.Json.Serialization.JsonNumberEnumConverter`1[TEnum]' violates the constraint of type parameter 'TEnum'.
>   at Program.\<Main\>$(String[] args)

### Regression?

Not a regression.  Both the `src` and `ref` changes came from <https://github.com/dotnet/runtime/pull/88984>, so they have never been consistent in any release.

### Known Workarounds

_No response_

### Configuration

.NET SDK 8.0.202\
.NET 8.0.3\
Windows 10 22H2 x64



### Other information

This bug might be considered part of <https://github.com/dotnet/runtime/issues/26187>.

The System.Enum type constraint is missing from the documentation page <https://learn.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonnumberenumconverter-1?view=net-8.0> as well.

The System.Text.Json 8.0.3 NuGet package does not carry any reference assemblies.  Thus, if you reference that and target .NET Framework, then you get the expected error CS0315 at compile time.

--- a/src/libraries/System.Text.Json/ref/System.Text.Json.cs
+++ b/src/libraries/System.Text.Json/ref/System.Text.Json.cs
@@ -978,7 +978,7 @@ namespace System.Text.Json.Serialization
         KebabCaseLower = 4,
         KebabCaseUpper = 5,
     }
-    public sealed partial class JsonNumberEnumConverter<TEnum> : System.Text.Json.Serialization.JsonConverterFactory where TEnum : struct
+    public sealed partial class JsonNumberEnumConverter<TEnum> : System.Text.Json.Serialization.JsonConverterFactory where TEnum : struct, System.Enum
     {
         public JsonNumberEnumConverter() { }
         public override bool CanConvert(System.Type typeToConvert) { throw null; }

